keps/0035-json-schema-export.md

```diff
@@ -0,0 +1,265 @@
+
+### User Stories
+
+#### Story 1
+
+A user wants to generate a UI to install a KUDO operator. He wants to use an automatic generator, for example https://rjsf-team.github.io/react-jsonschema-form/
+
+### Implementation Details
+
+#### Additional Types: 
+
+To bring the parameters up to par with JSON-schema, we add these additional types:
+- `boolean` A "true" or "false" value, from the JSON "true" or "false" productions
+- `number` An arbitrary-precision, base-10 decimal number value, from the JSON "number" production
+
+#### Additional Attributes
+
+JSON-Schema already provides good support for generated UIs, but it can be enhanced. Additionally, we do not want to adopt the full nested structure with alternatives and references yet, there is a different KEP for that. The additional attributes should provide a step in the direction to generate a good UI from it
+
+All of the additional attributes are fully optional.
+
+`group` 
+An attribute of type "string". This attribute describes a top-level group used for this parameter.
+
+`advanced` 
+An attribute of type "boolean" that indicates that this parameter is for advanced configuration and can be hidden by default. Defaults to "false". Parameters that have "advanced=true" *must* have a default value, as they might not be visible immediately.
+
+`hint`
+An attribute of type "string" that should provide a short description about the syntax of this parameter. Usually shown below the actual input element.
+
+`enum` 
+An array of values that are allowed to be used. Each entry must be of the type defined by the `type` attribute. If a default is specified, it must be listed as well.
+
+
+Example:
+
+```yaml
+  - name: NODE_DOCKER_IMAGE_PULL_POLICY
+    title: "Docker Image Pull Policy"
+    hint: "Cassandra node Docker image pull policy"
+    description: "The Docker image pull policy. Can be changed, for example if the image is pre-loaded on the Kubernetes nodes."
+    default: "Always"
+    type: "string"
+    enum: 
+      - "Always"
+      - "Never"
+      - "IfNotPresent"
+    group: "Node"
+    advanced: true
+    required: true
+```
+
+#### Groups
+
+Add a top-level element `groups` that describes the groups used:
+
+```
+apiVersion: kudo.dev/v1beta1
+groups:
+  - name: "Node"
+    title: "Node Configuration"
+    description: "This section contains parameters that are related to the Node"
+parameters:
+...
+```
+
+This `groups` element is fully optional and does not affect any behaviour of KUDO. We may decide to not add this to the OperatorVersion CRD but to keep it only in the `params.yaml` package format, that is an implementation detail. The data provided in this element can be used by the JSON-Schema export to fill in the metadata (title and description) of top level groups.
+
+### JSON-Schema export
+
+Extend the `kudo package list parameters` with an option `--output json-schema` to describe the parameters as JSON-Schema.
```

Comment:
I like this option

Reply:
I wonder how we can output a JSON-schema in YAML with this way though... `--output yaml-schema` is not really correct, `--output json-schema-as-yaml` might work, but is quite long...

And it might be nice to have the output as yaml, as the KEP-33 will support a JSON-Schema defined in a yaml file, so it would help with the conversion, and yaml is more common in the k8s world. But we can also add that later if required.